text/2021-09-27-refine-error-handling.md

```diff
@@ -0,0 +1,412 @@
+   public variable to store the error, and wrap it with `errors.WithStack` when
+   returning the error.
+
+   Example:
+
+   ```go
+   var (
+      ErrPodNotFound = errors.New("pod not found")
+
+      ErrPodNotRunning = errors.New("pod not running")
+   )
+
+   func handle() error {
+      return errors.WithStack(ErrPodNotFound)
+   }
+   ```
+
+3. Combine of rule 1 and rule 2. The context information is needed by the
+   caller, but it's also widely used by multiple functions. The variable name
+   could be more detailed than the type.
+
+   Example:
+
+   ```go
+   type ContainerRuntimeClientConnectErr struct {
+      ContainerRuntime string
+   }
+
+   var DockerConnectErr = ContainerRuntimeClientConnectErr {
+      ContainerRuntime: "docker"
+   }
+   ```
+
+4. Is this an error which will not be detected but appears in a lot of
+   functions? If so, you should use a global private variable, and wrap it with
+   `errors.WithStack`. If you want to share the same error (like `Not Found`) in
+   multiple packages, but the callers will never need to detect the `Not Found`,
+   please create a new `notFound` error under every package.
+
+   ```go
+   var errNotFound = errors.New("not found")
+   ```
+
+5. Is this really a simple error that will not appear in other functions? If so,
+   you should use an inline `errors.New`. The `errors.New` in `pkg/errors` is
+   already equipped with a stack backtrace, so you don't need to add it again.
+
+   Example:
+
+   ```go
+   func open(file string) error {
+      return errors.New("not found")
+   }
+   ```
+
+6. Are you propagating an error returned by other functions? If it's returned by
+   function inside the Chaos Mesh, we could assume this error is already
+   equipped with a stack backtrace, so there is not need to call
+   `errors.WithStack`. However, if it's returned by other libraries, it should
+   to call `errors.WithStack` to equip it with a stack backtrace. For more
+   information, see the section on error wrapping.
+
+   ```go
+   func startProcess(cmd *exec.Cmd) error {
+      err := cmd.Start()
+      if err != nil {
+         return nil, errors.WithStack(err)
+      }
+
+      return nil
+   }
+   ```
+
+### Error Wrapping
+
+* Return the original error if there is no additional context to add. If the
+  original error is not equipped with a stack, return with `errors.WithStack`.
+* Add context using
+  [`"pkg/errors".Wrap`](https://pkg.go.dev/github.com/pkg/errors#Wrap) so that
+  the error message provides more context
+
+   ```go
+   var ErrNotFound = errors.New("not found")
+
+   func open(file string) error {
+      return errors.Wrapf(ErrNotFound, "open file %s", file)
+   }
+   ```
+
+* Use [`"pkg/errors".Errorf`](https://pkg.go.dev/github.com/pkg/errors#Errorf)
+  with if the callers do not need to detect or handle that specific error case.
+
+The [`"pkg/errors".Wrap`](https://pkg.go.dev/github.com/pkg/errors#Wrap) will
```

Comment:
The line that appears here is very strange. Should we put this description at L#145?

Reply:
It actually applies for both entries above. I will modify the description.